app/controllers/base_opinion_controller.rb

```diff
@@ -0,0 +1,34 @@
+class BaseOpinionController < ApplicationController
+  before_action :check_for_opinion_collision, only: :create
+
+  def create
+    @opinion = current_user.send(opinion_type).new(permitted_params)
```

Comment:
ruby has `public_send` which expects to be used in first place in order to avoid invocation of private methods